apps/conv_layer/conv_layer_generator.cpp

```diff
@@ -0,0 +1,97 @@
+#include "Halide.h"
+
+namespace {
+
+using namespace Halide;
+
+class ConvolutionLayer : public Halide::Generator<ConvolutionLayer> {
+public:
+    GeneratorParam<bool>  auto_schedule{"auto_schedule", false};
+
+    ImageParam            input{Float(32), 4, "input"};
+    ImageParam            filter{Float(32), 4, "filter"};
+    ImageParam            bias{Float(32), 1, "bias"};
+
+    Func build() {
+        /* THE ALGORITHM */
+
+        Var x("x"), y("y"), z("z"), n("n");
+
+        Func f_conv("conv");
+        RDom r(0, 3, 0, 3, 0, 64);
+
+        f_conv(x, y, z, n) = bias(z);
+
+        f_conv(x, y, z, n) += filter(r.x, r.y, r.z, z) * input(x + r.x, y + r.y, r.z, n);
+
+        Func f_ReLU("ReLU");
+        f_ReLU(x, y, z, n) = max(0, f_conv(x, y, z, n));
+
+        /* THE SCHEDULE */
+
+        if (auto_schedule) {
+            // Provide estimates on the input image
+            input.dim(0).set_bounds_estimate(0, 131);
+            input.dim(1).set_bounds_estimate(0, 131);
+            input.dim(2).set_bounds_estimate(0, 64);
+            input.dim(3).set_bounds_estimate(0, 4);
+
+            filter.dim(0).set_bounds_estimate(0, 3);
+            filter.dim(1).set_bounds_estimate(0, 3);
+            filter.dim(2).set_bounds_estimate(0, 64);
+            filter.dim(3).set_bounds_estimate(0, 64);
+
+            bias.dim(0).set_bounds_estimate(0, 64);
+
+            // Provide estimates on the pipeline output
+            f_ReLU.estimate(x, 0, 128)
+                .estimate(y, 0, 128)
+                .estimate(z, 0, 64)
+                .estimate(n, 0, 4);
+
+            // Auto-schedule the pipeline
+            Pipeline p(f_ReLU);
+            p.auto_schedule(target);
+
+        } else if (get_target().has_gpu_feature()) {
+            Var ni, no, xi, xo, yi, yo, zi, zo;
+            f_ReLU.compute_root()
+                .split(x, xo, xi, 8)
+                .split(y, yo, yi, 8)
+                .split(z, zo, zi, 4)
+                .reorder(xi, yi, zi, n, xo, yo, zo)
+                .gpu_threads(xi, yi, zi)
+                .gpu_blocks(xo, yo, zo);
+
+            f_conv.compute_at(f_ReLU, n)
+                .gpu_threads(x, y, z)
+                .update()
+                .unroll(r.x)
+                .unroll(r.y)
+                .gpu_threads(x, y, z);
+        } else {
+            // Blocking spatially with vectorization
+            Var z_t("z_t"), y_t("y_t"), par("par");
+            int vec_len = 8;
+            int o_block_size = 32;
+            int y_block = 32;
+            f_conv.compute_root();
+            f_conv.fuse(z, n, par).parallel(par);
+            f_conv.update().reorder(x, y, r.z);
```

Comment:
The repetition of f_conv.update() is not idiomatic. Better to chain the schedule across multiple lines like above.